graph.c

```diff
@@ -0,0 +1,144 @@
+//idea: https://runestone.academy/ns/books/published/pythonds3/Graphs/AnAdjacencyList.html
+//Implementation of simple directed cycled weighted graph
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <glib.h>
+
+#define GRAPH_SIZE 5
+#define GRAPH_NEIGHBORS_TABLE_SIZE 5
+
+typedef struct {
+    int index;
+    int weight;
+} Graph_vertice_neighbor;
+
+typedef struct {
+    int index;
+    GArray *neigbors;
+} Graph_vertice;
+
+
+typedef struct {
+    GHashTable *vertices_table;
+} Graph;
+
+Graph graph_create();
+void graph_free(Graph *);
+void graph_list(Graph *);
+static void print_key(gpointer, gpointer, gpointer);
+
+void graph_add_vertice(Graph *, int);
+bool graph_erase_vertice(Graph *, int);
+Graph_vertice* graph_get_vertice(int);
+
+void graph_add_edge(Graph *, int, int, int);
+void graph_erase_edge(Graph *, int, int);
+
+static void print_key(gpointer key, gpointer value, gpointer user_data) {
+    printf("Key: %d\n", GPOINTER_TO_INT(key));
```

Comment:
@Yanpas @Dekz120
There is an output like:
```
Key: 1827303204
Key: 1827303204
Key: 1827303204
Key: 1827303204
```
but should be:
```
Key: 0
Key: 1
Key: 2
Key: 4
```

I try to use get_keys or get_keys_as_array from [glib](https://docs.gtk.org/glib/type_func.HashTable.foreach.html) but it is the same.
Keys are hashed but probably it is not hash.

Reply:
Consider using some IDE to expand macros (you are printing addresses instead of integers)

Reply:
https://clang.llvm.org/docs/AddressSanitizer.html

Reply:
A good catch!
I also used:
1. `clang --analyze`, it shows nothing about use-after-free although shows some interesting warnings.
2. `free(ptr)` and `ptr = NULL` in the end of `graph_add_vertice`

Reply:
> A good catch! I also use now:
> 
> 1. `clang --analyze`, it shows nothing about use-after-free although shows some interesting warnings.
> 2. `free(ptr)` and `ptr = NULL` in the end of `graph_add_vertice`

Reply:
I try to use lookup to compare index by value not by address

Reply:
Try to create index on heap to free the pointer in the end of function

Reply:
The problem solved by allocate int pointer in heap and assign index value to dereferenced pointer